### PR TITLE
Added hook for Item.isInCreativeTab

### PR DESCRIPTION
**Pull request by Cadiboo**

### CODE REVIEW

**Review by liach — COMMENTED** [review]
I recall there is another method that actually adds an item stack to the creative tab. You should patch that method instead.

**Comment by Cloudhunter** [comment]
What's the use case for this? PR's aren't generally accepted without one.

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 1]
```diff
@@ -58,17 +58,27 @@
...
  
-@@ -409,6 +412,9 @@
+@@ -409,8 +412,18 @@
  
      protected boolean func_194125_a(CreativeTabs p_194125_1_)
      {
 +        for (CreativeTabs tab : this.getCreativeTabs())
 +            if (tab == p_194125_1_)
 +                return true;
++
```
Comment (TechnicianLP) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: unnecessary newline

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 2]
```diff
@@ -58,17 +58,27 @@
...
          CreativeTabs creativetabs = this.func_77640_w();
-         return creativetabs != null && (p_194125_1_ == CreativeTabs.field_78027_g || p_194125_1_ == creativetabs);
+-        return creativetabs != null && (p_194125_1_ == CreativeTabs.field_78027_g || p_194125_1_ == creativetabs);
++        if (creativetabs != null) {
```
Comment (TechnicianLP) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: { mising newline

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 3]
```diff
@@ -58,17 +58,27 @@
...
++            if (p_194125_1_ == CreativeTabs.field_78027_g || p_194125_1_ == creativetabs)
++                return true;
++            if (net.minecraftforge.common.ForgeHooks.isInCreativeTab(this, p_194125_1_))
```
Comment (TechnicianLP) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: only supports adding to tabs not removing

File: src/main/java/net/minecraftforge/event/item/ItemEvent.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * ItemEvent is fired when an event involving any Item occurs.<br>
+ * If a method utilizes this {@link Item} as its parameter, the method will
+ * receive every child event of this class.<br>
+ * <br>
+ * {@link #item} contains the entity that caused this event to occur.<br>
+ * <br>
+ * All children of this event are fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+ **/
+public class ItemEvent extends Event
```
Comment (TechnicianLP) [inline src/main/java/net/minecraftforge/event/item/ItemEvent.java]: classname clashes with `net.minecraftforge.event.entity.item.ItemEvent` which is used for Events regarding `EntityItem`

File: src/main/java/net/minecraftforge/event/item/ItemEvent.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    public ItemEvent(Item item) {
+        this.item = item;
+    }
+
+    public Item getItem() {
+        return item;
+    }
+
+    @HasResult
+    public static class InCreativeTab extends ItemEvent {
```
Comment (TechnicianLP) [inline src/main/java/net/minecraftforge/event/item/ItemEvent.java]: should probably be made a standalone class as there currently is no other usecase for a `Item`-based `ItemEvent` class

**Review by TechnicianLP — CHANGES_REQUESTED** [review]
Interesting (yet probably vey rarely used) hook that breaks as soon as someone overrides that method.
Yet this hook only supports artificially adding items to creativetabs not removing them.

The already existing hook could probably be modified to be handled in `ForgeHooks` with that functionality added, while that would work it is questionable whether there is a better approach to this.

The `ItemEvent` class needs be renamed to not have a name-collision or removed altogether as there currently is no other usecase for it.

A PR should always include a description containing information not exclusively about why the hook is needed and whom it would benefit

(Yes i had time to waste)

File: src/main/java/net/minecraftforge/debug/item/ItemInCreativeTabTest.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @SubscribeEvent
+    public static void onInCreativeTab(final ItemEvent.InCreativeTab event) {
+        if (event.getCreativeTab().equals(CreativeTabs.INVENTORY)) {
+//            logger.info("not making Item " + event.getItem().getUnlocalizedName() + " appear in CreativeTab " + event.getCreativeTab().getTabLabel() + " because the creative tab \"inventory\" is just the Destroy Item button & should not have anything displayed in it");
+            return;
+        }
+
+        if (event.getItem().getRegistryName().getResourcePath().contains("z")) {
```
Comment (TechnicianLP) [inline src/main/java/net/minecraftforge/debug/item/ItemInCreativeTabTest.java]: this should probably be changed to specific Items to make it clear which Items are affected (currently BlazeRod, BlazeDust and Quartz)

File: patches/minecraft/net/minecraft/item/Item.java.patch (hunk 2 above)
Comment (Daomephsta) [reply patches/minecraft/net/minecraft/item/Item.java.patch]: Patches don't follow the same style as Forge classes. They are supposed to be minimal, so braces don't get a new line.

**Comment by Cadiboo** [comment]
> What's the use case for this? PR's aren't generally accepted without one.

The use case for this was something pretty small - It could have been fixed by using the CreativeTabs.displayAllRelevantItems method, but I forgot all about it.

I went about making this hook the wrong way, but I think it would still be useful to be able to stop items being displayed in creative tabs without having to override the item.

**Pull request closed without merging**